### PR TITLE
[Traceable FSDP2][Inductor] Support non-functionalized fsdp.split_with_sizes_copy; enable inductor E2E unit test for simple MLP

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129422

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -5297,6 +5297,12 @@ def __init__(
             self.mutation_names.append(tensor_args[0].get_name())
             return
 
+        if self.op_overload is torch.ops.fsdp.split_with_sizes_copy.default:
```
Comment (eellison) [inline torch/_inductor/ir.py:5300]: How does this node get inserted in the first place ?

I was going to say something along the lines of generalize this mutation analysis but it doesn't seem like this will occur very often:

```
# The unsupported cases usually do not show up here (because
# AOTAutograd functionalized them away); the only way for an in-place
# op to show up here is if a lowering or pass introduced it.```

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -292,10 +292,6 @@ def wrapped(*args, **kwargs):
             unpacked = True
             args = args[0]
 
-        # explicitly assert for "out=" ops for better error messages
```
Comment (eellison) [inline torch/_inductor/lowering.py:295]: Is the metadata mutated at all here ? I'm not sure we handle that, maybe we can check ?

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -292,10 +292,6 @@ def wrapped(*args, **kwargs):
...
-        assert not any(
-            x == "out" for x in kwargs.keys()
```
Comment (eellison) [inline torch/_inductor/lowering.py:297]: We should either :

1) Generalize the fallback kernel `out` mutation handling (only handles fsdp.split_with_sizes_copy currently)

2) Still error on other ops with `kwargs=out`

**Review by eellison — COMMENTED** [review]
looks good, a few comments

File: torch/_inductor/ir.py (hunk 1 above)
Comment (yf225) [reply torch/_inductor/ir.py:5300]: yes we intentionally preserve this op through the lowering stack (no functionalization, no decomposition): https://github.com/pytorch/pytorch/pull/129414/files: no functionalization to more easily work around AOTAutograd limitations, no decomp so that we can do pattern-matching on it much more easily (in future PR).

would love to make this more general - maybe I should create a dictionary where each op maps to its specific mutation analysis? (although there will only be one op `fsdp.split_with_sizes_copy` for now)

File: torch/_inductor/lowering.py (hunk 3 above)
Comment (yf225) [reply torch/_inductor/lowering.py:297]: yes I think I should just generalize the fallback kernel `out` mutation handling to cover `out=Tensor` and `out=List[Tensor]` to start

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (yf225) [reply torch/_inductor/lowering.py:295]: agreed - wonder what would be the best way to check it? (do we have something like `example_value` where we can e.g. compare the before vs. after strides for the `out` arg?)

File: torch/_inductor/ir.py (hunk 1 above)
Comment (zou3519) [reply torch/_inductor/ir.py]: I'm confused about the strategy here. How do we ensure there is no silent incorrectness if we preserve this operator through the PT2 stack without functionalization? I would have expected that we auto-functionalize it and then later re-inplace it.

Comment (bdhirsh) [reply torch/_inductor/ir.py]: Let me give some background first: Will was running into an issue in fsdp where he had a graph (from compiled autograd) with the following traits:

(1) two inputs of the graph were aliased to each other
(2) one of them is mutated (a resize_ and set_ to handle the all-gathered param)
(3) they are both subclasses

This situation is "handled" in AOTAutograd today, but (a) not in a very performant way (we emit `as_strided` and `as_strided_scatter` in several places) and (b) we error on subclasses, so this doesn't work at all for FSDP.

The compromise we were looking into was "de-functionalization", for a small number of mutable ops that only FSDP cares about: `resize_` and `set_`. This "solves" the problem above, because from AOTAutograd pov there are no mutations that functionalization needs to handle. "Not functionalizing" also has some requirements though:

(1) we need to not DCE the mutable op
(2) inductor needs to handle this mutable op directly in its IR (previously, mutable ops could only show up as `copy_` or `set_` nodes, and only showed up in a small epilogue at the end of the graph)

I think we handle (1) properly today, just by adding the two specific ops to FX's allow-list. I'm not sure what the status of (2) is (although this change is probably trying to improve it).

For `fsdp.split_with_sizes_copy.out`: this is the 3rd "interesting" mutable op that FSDP cares about. @yf225 will remember better, but if this op gets invoked on one of the aliased inputs, then I think we'll need to de-functionalize it to avoid the AOT issue mentioned above. If this op is not actually invoked on either of the aliased inputs, then in theory we could do the normal auto-functionalize flow for custom ops on it. Although the reinplacing pass will be load-bearing for FSDP, so we probably will want some tests that reinplacing always works properly on that op.

@zou3519 let me know what you think

Comment (bdhirsh) [reply torch/_inductor/ir.py]: my other baked-in assumption: if we are already de-functionalizing `set_` and `resize_` that already requires a lift from inductor to handle mutable ops in the middle of the graph, so "one extra mutable op" isn't too much of a lift. Does that assumption sound wrong?

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/lowering.py:295]: After running an `out kernel` you can go iterate through the `args_flat` and check they have the same metadata as the equivalent arg in `tensor_args`

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L4171

File: torch/_inductor/ir.py (hunk 1 above)
Comment (zou3519) [reply torch/_inductor/ir.py]: My take is that morally, "de-functionalizing" should be a last resort (otherwise, why not de-functionalize *all* custom ops?). Since `fsdp.split_with_sizes_copy.out` looks like a standard op from the outside, we should functionalize it via our standard mechanism (auto_functionalize).

I expected "de-functionalizing" to cause some sort of barrier mechanism in inductor that would be harmful to performance, hence why it should be a last resort.
(for (2) -- naively I would have expected that we partition the graph based on the set_, pass each functional subgraph to inductor's post_grad passes, and then stitch the graph back together, which means there is a performance implication for not being able to optimize across `set_` calls).

Side note: I want to rename "de-functionalizing" -- it's more like "forgoing functionalization"

Comment (zou3519) [reply torch/_inductor/ir.py]: > For fsdp.split_with_sizes_copy.out: this is the 3rd "interesting" mutable op that FSDP cares about. @yf225 will remember better, but if this op gets invoked on one of the aliased inputs, then I think we'll need to de-functionalize it to avoid the AOT issue mentioned above

I don't fully understand the problem. Is the claim that auto_functionalization will do something subpar, so we need to not do it?

Comment (yf225) [reply torch/_inductor/ir.py]: Discussed with @zou3519 offline, the details are:

It might look like `fsdp.split_with_sizes_copy` is a standard op so auto-functionalize should just work. But the reality is that `fsdp.split_with_sizes_copy` is (transitively) mutating a graph input but auto-functionalize doesn't know about it, and hence it will not issue the `.copy_` to copy the output back to the original tensor to preserve semantics:
```python
def f(x, y):
    tmp = torch.empty_like(out)
    res = auto_functionalized(op, x, out=tmp)
    # we need a tmp.copy_(res) here, but we don't have it
    y.set_(tmp)
```

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 4]
```diff
@@ -86,7 +86,7 @@ def all_gather_copy_in_cuda(
 
 
 lib.define(
-    "split_with_sizes_copy(Tensor all_gather_output, SymInt[] all_gather_input_split_sizes, int dim=0, *, Tensor(a!)[] out) -> ()"
+    "split_with_sizes_copy(Tensor all_gather_output, SymInt[] all_gather_input_split_sizes, int dim=0, *, Tensor(a!)[] ret) -> ()"
```
Comment (yf225) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:89]: @zou3519 renaming `out` to `ret` here works around the `out` check in Inductor - but wonder should we actually do it (what would be the difference between `out` support and `ret` support?)

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -5470,7 +5486,6 @@ def get_inputs_that_alias_output(self):
         return self.alias_names
 
     def get_mutation_names(self):
-        assert len(self.mutation_names) <= 1
```
Comment (zou3519) [inline torch/_inductor/ir.py:5473]: Based on https://github.com/pytorch/pytorch/pull/112547, we should not be returning more than one name in `get_mutation_names`. Instead we may need to go the route that triton kernels did for mutating multiple inputs? @oulgen @Chillee can you check me on this?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 4 above)
Comment (zou3519) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:89]: Let's put `out` back here, we're trying to fix the general problem. By the way, the out= assertion is actually wrong. We should assert against `out` being a Tensor ONLY IF the operator is not a fallback. This will get us past the `out=` assertion.

Code archaeology got me to https://github.com/pytorch/pytorch/pull/88367, and it looks like we just wanted to make the error message of the following assert better (which only triggers if the operator is not a fallback)

**Review by zou3519 — COMMENTED** [review]
We should rename `ret` back into `out` and squash out the problems. There's a bug in the `out=` assertion (fixing that will get us past it). Then there's the question of how to handle an operator with multiple mutations -- it looks like there is prior art from the triton kernel work so maybe Oguz or Horace can chip in here.

File: torch/_inductor/ir.py (hunk 5 above)
Comment (oulgen) [reply torch/_inductor/ir.py:5473]: Yep, inductor does not support more than 1 mutation names, you'll hit another assert in scheduler. You can do the same thing I did for triton kernels

File: torch/_inductor/scheduler.py [hunk 6]
```diff
@@ -1656,8 +1656,6 @@ def add_user(
             else:
                 node_mode = None
 
-            # a node will mutate either 0 or 1 buffers
-            assert len(node.get_mutations()) <= 1
```
Comment (oulgen) [inline torch/_inductor/scheduler.py:1660]: this is incorrect, you'll end up overwriting data right below

File: torch/_inductor/ir.py (hunk 5 above)
Comment (yf225) [reply torch/_inductor/ir.py:5615]: Changed `FallbackKernel` to use same approach as `UserDefinedTritonKernel`. This is changing existing code and will see if any unit test error

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (yf225) [reply torch/_inductor/lowering.py:295]: Thanks! added the check in https://github.com/pytorch/pytorch/pull/129422/files#diff-cf6ca00beddc32a2a6a2933fb9913b6a2b925ffc3b745488967210e4343134acR4328

**Comment by yf225** [comment]
Updated PR based on review comments @zou3519 @oulgen @eellison @bdhirsh

File: torch/_inductor/ir.py (hunk 1 above)
Comment (yf225) [reply torch/_inductor/ir.py]: Updated the PR to go with the original direction: not functionalizing `fsdp.split_with_sizes_copy`.

File: torch/_functorch/_aot_autograd/functional_utils.py [hunk 7]
```diff
@@ -26,6 +26,10 @@
 
 aot_joint_log = getArtifactLogger(__name__, "aot_joint_graph")
 
+# These are mutation ops that can show up in the middle of the graph,
+# because they are ops that we explicitly do **not** functionalize
+_avoid_functionalize_ops: Set[Callable] = set()
```
Comment (zou3519) [inline torch/_functorch/_aot_autograd/functional_utils.py:31]: Where are we adding set_ / resize_?

File: torch/_inductor/ir.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_inductor/ir.py]: I repro'd the silent correctness locally in that test when running with `aot_eager`: I think we can keep using auto-functionalize for `split_with_sizes_copy`, but there was a bug in the "do-not-functionalize" logic for set_. This patch fixed the test for me:
```
diff --git a/torch/distributed/_composable/fsdp/_fsdp_param.py b/torch/distributed/_composable/fsdp/_fsdp_param.py
index 955beb86699..866108030d0 100644
--- a/torch/distributed/_composable/fsdp/_fsdp_param.py
+++ b/torch/distributed/_composable/fsdp/_fsdp_param.py
@@ -76,6 +76,8 @@ def set_(tensor, data):

 @torch.library.impl(lib, "set_", "Functionalize")
 def set__functionalize(tensor, data):
+    torch._sync(tensor)
+    torch._sync(data)
     tensor_inner = torch._from_functional_tensor(tensor)
     data_inner = torch._from_functional_tensor(data)
     tensor_inner.set_(data_inner)  # type: ignore[call-overload]
```

I ended up making a minimal repro below that shows the same correctness problem. This issue is that when we dump `set_` into the otherwise-functional graph, we need to make sure that its inputs have had any pending mutations properly "flushed". Otherwise, the inputs that we pass into the `set_` will be the values of those inputs before any mutations were performed on them.
```
import torch
import numpy as np

import torch.distributed._composable.fsdp._fsdp_param

@torch.library.custom_op("mylib::add_one_out", mutates_args={"out"})
def add_one_out(x: torch.Tensor, out: torch.Tensor) -> None:
    torch.add(x, 1, out=out)

@torch.compile(backend="aot_eager")
def f(x):
    buf = torch.zeros(2)
    buf_view = buf.view(-1)
    torch.ops.mylib.add_one_out(x, out=buf_view)
    buf_view2 = buf.view(-1)
    torch.ops.fsdp.set_(x, buf_view2)

x = torch.zeros(2)
f(x)
print(x)
```

File: torch/_inductor/ir.py (hunk 5 above)
Comment (eellison) [reply torch/_inductor/ir.py:5615]: What is lacking in the support for > 1 mutation names ? It seems to generalize to multiple mutations:

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/scheduler.py#L1653-L1667

**Comment by yf225** [comment]
We fixed the issue in `fsdp.set_` functionalization and hence we can do auto-functionalize for `fsdp.split_with_sizes_copy` now. Hence closing this PR.

File: torch/_inductor/ir.py (hunk 1 above)
Comment (yf225) [reply torch/_inductor/ir.py]: As Brian mentioned, fixing the `set_` "do-not-functionalize" logic allows us to use auto-functionalize for `split_with_sizes_copy`. We took this approach in https://github.com/pytorch/pytorch/pull/129502 and were unblocked on enabling E2E inductor unit test.
